org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/CodeActionMarkerResolution.java

```diff
@@ -55,6 +64,18 @@ public String getLabel() {
 
 	@Override
 	public void run(IMarker marker) {
+		if (codeAction.getEdit() == null) {
+			String languageServerId = marker.getAttribute(LSPDiagnosticsToMarkers.LANGUAGE_SERVER_ID, (String)null);
+			LanguageServerDefinition definition = languageServerId != null ? LanguageServersRegistry.getInstance().getDefinition(languageServerId) : null;
+			try {
+				CompletableFuture<LanguageServer> ls = definition != null ? LanguageServiceAccessor.getInitializedLanguageServer(marker.getResource(), definition, cap -> true) : null;
```

Comment:
I wonder if we should add a method
	public static CompletableFuture<LanguageServer> getInitializedLanguageServer(@NonNull LanguageServerDefinition lsDefinition, Predicate<ServerCapabilities> capabilitiesPredicate)
			throws IOException
that does not need a IFile or IResource, in the end we already know which LS we need, so the resource is not really going to change the server we are looking for.

Reply:
> public static CompletableFuture getInitializedLanguageServer(@nonnull LanguageServerDefinition lsDefinition, Predicate capabilitiesPredicate)

There can be multiple instance of a language server for a given LanguageServerDefinition; not all language servers set "singleton" nor support workapceFolders. So the resource matters.